feanil/api docs (#32971)
openedx/edx-platform: docs/how-tos/use_the_api.rst

How To Use the REST API
#######################

.. How-tos should have a short introduction sentence that captures the user's goal and introduces the steps.

> This how-to will help get setup to be able to make aunthenticated requests to

Reviewer: [inform] You may want to review https://edx-enterprise-api.readthedocs.io/en/latest/authentication.html to see if there is any details from there you'd like to add. Note that @pomegranited asked elsewhere about token expiration, and this docs has something on that.

Author: Cross linked for now.

   how-tos if possible.

* You have access to the edx-platform Django Admin (``/admin``) Panel.

* You have a user that you want to make the rest calls as (``UserA``).

* You are familiar with `the basics of HTTP and Rest`_

* For the purposes of this tutorial we'll assume your LMS is located at
  https://lms.example.com

.. _the basics of HTTP and Rest: https://code.tutsplus.com/a-beginners-guide-to-http-and-rest--net-16340t

Steps
*****

.. A task should have 3 - 7 steps.  Tasks with more should be broken down into digestible chunks.

#. Go to https://lms.example.com/admin/oauth2_provider/application/

#. Click :guilabel:`Add Application`

#. Choose "UserA" for the user.

#. Choose ``Confidential`` Client Type

#. Choose "Client Credentials" for the Authorization Grant Type

#. Set a name for your application.

#. Save the ``client_id`` and ``client_secret``.

#. The best way to interact with the edx-platform REST API is by making
   requests using the JWT Authorization header.  User the ``client_id`` and
   ``client_secret`` to get a JWT token.

   .. code-block:: python

      import base64
      import requests

      client_id = "vovj0AItd9EnrOKjkDli0HpSF9HoooaTY9yueafn"

>       client_secret = "a3Fkwr24dfDSlIXt3v3q4Ob41CYQNZyGmtK8Y8ax0srpIa2vJON3OC5Rvj1i1wizsIUv1W1qM1Q2XPeuyjucNixsHXZsuw1dn2B9nH3IyjSvuFb5KoydDvWX8Hx8znqD"

Reviewer: Do we want a comment or change to the code to encourage users to better protect the secret?

      credential = f"{client_id}:{client_secret}"
      encoded_credential = base64.b64encode(credential.encode("utf-8")).decode("utf-8")

>       headers = {"Authorization": f"Basic {encoded_credential}", "Cache-Control": "no-cache"}

Reviewer: Note that I'm hoping to remove BasicAuthentication as a default, so this is a good reminder that there are some _potential_ uses on some endpoints.

Author: It looks like the relevant endpoints do have overrides so we can still use this with Client credentials even after we remove the `BasicAuthentication` default (tested locally.)

Reviewer: I'm curious how this actually works. I don't see any references to `BasicAuthentication` in https://github.com/search?q=repo%3Aopenedx%2Fedx-platform%20BasicAuthentication&type=code. Not sure if you'd want to look into this, but it feels like a gap in knowledge about how our authentication works, at least on my part.

>       data = {"grant_type": "client_credentials", "token_type": "jwt"}

Reviewer: In place of `BasicAuthentication`, I always used `client_id` and `client_secret` as params.

Author: I think using the `client_id` as a param with public clients is fine but using the `client_secret` in the query params exposes it to various HTTP logs and so it can leak more easily in production environments.  The Authorization header should be within the TLS transaction so it should be more secure.

Reviewer: I'm planning on removing `BasicAuthentication` as part of https://github.com/openedx/edx-platform/issues/33028. Do your instructions in-fact rely on `BasicAuthentication`? Should we make it an explicit auth class on `/access_token`, or does this work in some other way? As I said, I've never done this before so I wasn't aware it was an option.

Author: I dug into it, the reason it's working is because the views behind `/access_token` are just django views and not DRF views.  They go from oauth_dispatch to  DOT and don't use DRF to do manage permissions.

      token_request = requests.post(

>           "http://localhost:8000/oauth2/access_token", headers=headers, data=data

Reviewer: Do you want to be more consistent with `lms.example.com` or `localhost:8000`?

Author: Updated everything to use `lms.example.com`

      )
      access_token = token_request.json()["access_token"]

> #. The code above will produce a JWT token that you can use to hit any existing

Reviewer: This can only hit endpoints that accept JwtAuthentication.

Author: Yea, but that's complicated and we're working to fix it so I consider things that don't work with this to be a bug.

Reviewer: You might want a ticket for addressing any endpoints that are [explicitly setting authentication_classes](https://github.com/search?q=repo%3Aopenedx%2Fedx-platform%20authentication_classes&type=code) and are not including `JwtAuthentication`. See this [example view with this issue](https://github.com/openedx/edx-platform/blob/ddb092c07c6e2b663dc3d52e7197b42f7f017988/lms/djangoapps/badges/api/views.py#L97-L100).

Reviewer: @feanil: Feel free to mark resolved with or without ticketing any follow-up.

How To Use the REST API
#######################

.. How-tos should have a short introduction sentence that captures the user's goal and introduces the steps.

> This how-to will help get setup to be able to make aunthenticated requests to

Reviewer: ```suggestion
This how-to will help you get setup to be able to make authenticated requests to
```
Or even simpler:
```
This how-to will help you make authenticated requests to
```

the edx-platform REST API.

Assumptions
***********

.. This section should contain a bulleted list of assumptions you have of the
   person who is following the How-to.  The assumptions may link to other
   how-tos if possible.

* You have access to the edx-platform Django Admin (``/admin``) Panel.

* You have a user that you want to make the rest calls as (``UserA``).

* You are familiar with `the basics of HTTP and Rest`_

* For the purposes of this tutorial we'll assume your LMS is located at
  https://lms.example.com

.. _the basics of HTTP and Rest: https://code.tutsplus.com/a-beginners-guide-to-http-and-rest--net-16340t

Steps
*****

.. A task should have 3 - 7 steps.  Tasks with more should be broken down into digestible chunks.

#. Go to https://lms.example.com/admin/oauth2_provider/application/

#. Click :guilabel:`Add Application`

#. Choose "UserA" for the user.

#. Choose ``Confidential`` Client Type

#. Choose "Client Credentials" for the Authorization Grant Type

#. Set a name for your application.

> #. Save the ``client_id`` and ``client_secret``.

Reviewer: [inform] You don't need to make any changes now, but just letting you know that the client_secret will soon be viewable one-time, and would need to be moved to a secure location like a password manager.

#. The best way to interact with the edx-platform REST API is by making

>    requests using the JWT Authorization header.  User the ``client_id`` and

Reviewer: ```suggestion
   requests using the JWT Authorization header.  Use the ``client_id`` and
```

   ``client_secret`` to get a JWT token.

   .. code-block:: python

      import base64
      import requests

      client_id = "vovj0AItd9EnrOKjkDli0HpSF9HoooaTY9yueafn"

>       # Client secretes should not be exposed in your code, we put it here to

Reviewer: ```suggestion
      # Client secrets should not be exposed in your code, we put it here to
```
